File: tests/algorithms/x/Java/dynamic_programming/longest_common_subsequence.java
public class Main {
    static class LcsResult {
        int length;
        String sequence;
        LcsResult(int length, String sequence) {
            this.length = length;
            this.sequence = sequence;
        }
        LcsResult() {}
        @Override public String toString() {
            return String.format("{'length': %s, 'sequence': '%s'}", String.valueOf(length), String.valueOf(sequence));
        }
    }

    static String a;
    static String b;
    static LcsResult res;

    static int[][] zeros_matrix(int rows, int cols) {
        int[][] matrix = ((int[][])(new int[][]{}));
        int i = 0;
        while (i <= rows) {
            int[] row = ((int[])(new int[]{}));
            int j = 0;
            while (j <= cols) {
                row = ((int[])(java.util.stream.IntStream.concat(java.util.Arrays.stream(row), java.util.stream.IntStream.of(0)).toArray()));
                j = j + 1;
            }
            matrix = ((int[][])(appendObj((int[][])matrix, row)));
            i = i + 1;
        }
        return matrix;
    }

    static LcsResult longest_common_subsequence(String x, String y) {
        int m = _runeLen(x);
        int n = _runeLen(y);
        int[][] dp = ((int[][])(zeros_matrix(m, n)));
        int i_1 = 1;
        while (i_1 <= m) {
            int j_1 = 1;
            while (j_1 <= n) {
                if ((x.substring(i_1 - 1, i_1 - 1+1).equals(y.substring(j_1 - 1, j_1 - 1+1)))) {
dp[i_1][j_1] = dp[i_1 - 1][j_1 - 1] + 1;
                } else                 if (dp[i_1 - 1][j_1] > dp[i_1][j_1 - 1]) {
dp[i_1][j_1] = dp[i_1 - 1][j_1];
                } else {
dp[i_1][j_1] = dp[i_1][j_1 - 1];
                }
                j_1 = j_1 + 1;
            }
            i_1 = i_1 + 1;
        }
        String seq = "";
        int i2 = m;
        int j2 = n;
        while (i2 > 0 && j2 > 0) {
            if ((x.substring(i2 - 1, i2 - 1+1).equals(y.substring(j2 - 1, j2 - 1+1)))) {
                seq = x.substring(i2 - 1, i2 - 1+1) + seq;
                i2 = i2 - 1;
                j2 = j2 - 1;
            } else             if (dp[i2 - 1][j2] >= dp[i2][j2 - 1]) {
                i2 = i2 - 1;
            } else {
                j2 = j2 - 1;
            }
        }
        return new LcsResult(dp[m][n], seq);
    }
    public static void main(String[] args) {
        a = "AGGTAB";
        b = "GXTXAYB";
        res = longest_common_subsequence(a, b);
        System.out.println("len = " + _p(res.length) + ", sub-sequence = " + res.sequence);
    }

    static <T> T[] appendObj(T[] arr, T v) {
        T[] out = java.util.Arrays.copyOf(arr, arr.length + 1);
        out[arr.length] = v;
        return out;
    }

    static int _runeLen(String s) {
        return s.codePointCount(0, s.length());
    }

    static String _p(Object v) {
        if (v == null) return "<nil>";
        if (v.getClass().isArray()) {
            if (v instanceof int[]) return java.util.Arrays.toString((int[]) v);
            if (v instanceof long[]) return java.util.Arrays.toString((long[]) v);
            if (v instanceof double[]) return java.util.Arrays.toString((double[]) v);
            if (v instanceof boolean[]) return java.util.Arrays.toString((boolean[]) v);
            if (v instanceof byte[]) return java.util.Arrays.toString((byte[]) v);
            if (v instanceof char[]) return java.util.Arrays.toString((char[]) v);
            if (v instanceof short[]) return java.util.Arrays.toString((short[]) v);
            if (v instanceof float[]) return java.util.Arrays.toString((float[]) v);
            return java.util.Arrays.deepToString((Object[]) v);
        }
        return String.valueOf(v);
    }
}

File: tests/algorithms/x/Pascal/dynamic_programming/longest_common_subsequence.pas
{$mode objfpc}
program Main;
uses SysUtils;
type LcsResult = record
  length_: integer;
  sequence: string;
end;
type IntArray = array of integer;
type IntArrayArray = array of IntArray;
var _nowSeed: int64 = 0;
var _nowSeeded: boolean = false;
procedure init_now();
var s: string; v: int64;
begin
  s := GetEnvironmentVariable('MOCHI_NOW_SEED');
  if s <> '' then begin
    Val(s, v);
    _nowSeed := v;
    _nowSeeded := true;
  end;
end;
function _now(): integer;
begin
  if _nowSeeded then begin
    _nowSeed := (_nowSeed * 1664525 + 1013904223) mod 2147483647;
    _now := _nowSeed;
  end else begin
    _now := Integer(GetTickCount64()*1000);
  end;
end;
function _bench_now(): int64;
begin
  _bench_now := GetTickCount64()*1000;
end;
function _mem(): int64;
var h: TFPCHeapStatus;
begin
  h := GetFPCHeapStatus;
  _mem := h.CurrHeapUsed;
end;
procedure panic(msg: string);
begin
  writeln(msg);
  halt(1);
end;
var
  bench_start_0: integer;
  bench_dur_0: integer;
  bench_mem_0: int64;
  bench_memdiff_0: int64;
  a: string;
  b: string;
  res: LcsResult;
  y: string;
  rows: integer;
  x: string;
  cols: integer;
function makeLcsResult(length_: integer; sequence: string): LcsResult; forward;
function zeros_matrix(rows: integer; cols: integer): IntArrayArray; forward;
function longest_common_subsequence(x: string; y: string): LcsResult; forward;
function makeLcsResult(length_: integer; sequence: string): LcsResult;
begin
  Result.length_ := length_;
  Result.sequence := sequence;
end;
function zeros_matrix(rows: integer; cols: integer): IntArrayArray;
var
  zeros_matrix_matrix: array of IntArray;
  zeros_matrix_i: integer;
  zeros_matrix_row: array of integer;
  zeros_matrix_j: integer;
begin
  zeros_matrix_matrix := [];
  zeros_matrix_i := 0;
  while zeros_matrix_i <= rows do begin
  zeros_matrix_row := [];
  zeros_matrix_j := 0;
  while zeros_matrix_j <= cols do begin
  zeros_matrix_row := concat(zeros_matrix_row, IntArray([0]));
  zeros_matrix_j := zeros_matrix_j + 1;
end;
  zeros_matrix_matrix := concat(zeros_matrix_matrix, [zeros_matrix_row]);
  zeros_matrix_i := zeros_matrix_i + 1;
end;
  exit(zeros_matrix_matrix);
end;
function longest_common_subsequence(x: string; y: string): LcsResult;
var
  longest_common_subsequence_m: integer;
  longest_common_subsequence_n: integer;
  longest_common_subsequence_dp: IntArrayArray;
  longest_common_subsequence_i: integer;
  longest_common_subsequence_j: integer;
  longest_common_subsequence_seq: string;
  longest_common_subsequence_i2: integer;
  longest_common_subsequence_j2: integer;
begin
  longest_common_subsequence_m := Length(x);
  longest_common_subsequence_n := Length(y);
  longest_common_subsequence_dp := zeros_matrix(longest_common_subsequence_m, longest_common_subsequence_n);
  longest_common_subsequence_i := 1;
  while longest_common_subsequence_i <= longest_common_subsequence_m do begin
  longest_common_subsequence_j := 1;
  while longest_common_subsequence_j <= longest_common_subsequence_n do begin
  if x[longest_common_subsequence_i - 1+1] = y[longest_common_subsequence_j - 1+1] then begin
  longest_common_subsequence_dp[longest_common_subsequence_i][longest_common_subsequence_j] := longest_common_subsequence_dp[longest_common_subsequence_i - 1][longest_common_subsequence_j - 1] + 1;
end else begin
  if longest_common_subsequence_dp[longest_common_subsequence_i - 1][longest_common_subsequence_j] > longest_common_subsequence_dp[longest_common_subsequence_i][longest_common_subsequence_j - 1] then begin
  longest_common_subsequence_dp[longest_common_subsequence_i][longest_common_subsequence_j] := longest_common_subsequence_dp[longest_common_subsequence_i - 1][longest_common_subsequence_j];
end else begin
  longest_common_subsequence_dp[longest_common_subsequence_i][longest_common_subsequence_j] := longest_common_subsequence_dp[longest_common_subsequence_i][longest_common_subsequence_j - 1];
end;
end;
  longest_common_subsequence_j := longest_common_subsequence_j + 1;
end;
  longest_common_subsequence_i := longest_common_subsequence_i + 1;
end;
  longest_common_subsequence_seq := '';
  longest_common_subsequence_i2 := longest_common_subsequence_m;
  longest_common_subsequence_j2 := longest_common_subsequence_n;
  while (longest_common_subsequence_i2 > 0) and (longest_common_subsequence_j2 > 0) do begin
  if x[longest_common_subsequence_i2 - 1+1] = y[longest_common_subsequence_j2 - 1+1] then begin
  longest_common_subsequence_seq := x[longest_common_subsequence_i2 - 1+1] + longest_common_subsequence_seq;
  longest_common_subsequence_i2 := longest_common_subsequence_i2 - 1;
  longest_common_subsequence_j2 := longest_common_subsequence_j2 - 1;
end else begin
  if longest_common_subsequence_dp[longest_common_subsequence_i2 - 1][longest_common_subsequence_j2] >= longest_common_subsequence_dp[longest_common_subsequence_i2][longest_common_subsequence_j2 - 1] then begin
  longest_common_subsequence_i2 := longest_common_subsequence_i2 - 1;
end else begin
  longest_common_subsequence_j2 := longest_common_subsequence_j2 - 1;
end;
end;
end;
  exit(makeLcsResult(longest_common_subsequence_dp[longest_common_subsequence_m][longest_common_subsequence_n], longest_common_subsequence_seq));
end;
begin
  init_now();
  bench_mem_0 := _mem();
  bench_start_0 := _bench_now();
  a := 'AGGTAB';
  b := 'GXTXAYB';
  res := longest_common_subsequence(a, b);
  writeln((('len = ' + IntToStr(res.length_)) + ', sub-sequence = ') + res.sequence);
  bench_memdiff_0 := _mem() - bench_mem_0;
  bench_dur_0 := (_bench_now() - bench_start_0) div 1000;
  writeln('{');
  writeln(('  "duration_us": ' + IntToStr(bench_dur_0)) + ',');
  writeln(('  "memory_bytes": ' + IntToStr(bench_memdiff_0)) + ',');
  writeln(('  "name": "' + 'main') + '"');
  writeln('}');
end.
